Mindtpy rewrite

ZedongPeng opened the pull request:
## Fixes # .

## Summary/Motivation:


## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

bernalde commented on pyomo/contrib/mindtpy/algorithm_base_class.py:
```diff
@@ -0,0 +1,2574 @@
+        correctly set up the config arguments and initialize the generic parts
+        of the algorithm state.
+
+        这里是应该包含所有set_up_solve_data中的数据
```
Please remove these comments in Chinese or translate them to English

bernalde commented on pyomo/contrib/mindtpy/algorithm_base_class.py:
```diff
@@ -0,0 +1,2574 @@
+        self.mip_iter = 0
+        self.mip_subiter = 0
+        self.nlp_infeasible_counter = 0
+        # TODO: 这个init_strategy的判断应该不需要吧
```
Please remove these comments in Chinese or translate them to English

bernalde commented on pyomo/contrib/mindtpy/algorithm_base_class.py:
```diff
@@ -0,0 +1,2574 @@
+        self.best_solution_found = None
+        self.best_solution_found_time = None
+
+        # 下面的是跟cofig相关的，上面已经定义config了，那是不是放在这里也可以？
```
Please remove these comments in Chinese or translate them to English

bernalde commented on pyomo/contrib/mindtpy/algorithm_base_class.py:
```diff
@@ -0,0 +1,2574 @@
+        # 下面的很多都是根据具体输入模型相关的，不应该放到__init__中
+        # 这里config看一下是否有一些个性化的配置
+        # if the objective function is a constant, dual bound constraint is not added.
+        # TODO: 这个应该可以换一个位置
```
Please remove these comments in Chinese or translate them to English

bernalde commented on pyomo/contrib/mindtpy/algorithm_base_class.py:
```diff
@@ -0,0 +1,2574 @@
+        # # -----------------------
+
+        # self.primal_bound_progress = [self.primal_bound]
+        # self.dual_bound_progress = [self.dual_bound]
```
If unused and unnecessary, remove this commented code.

bernalde commented on pyomo/contrib/mindtpy/algorithm_base_class.py:
```diff
@@ -0,0 +1,2574 @@
+                    # TODO: rewrite 
+                    results = self._gather_problem_info_and_solve_non_gdps(
+                        model, config)
+                    # If it wasn't disjunctive, we solved it
```
I think this no longer applies, right?

bernalde commented on pyomo/contrib/mindtpy/algorithm_base_class.py:
```diff
@@ -0,0 +1,2574 @@
+                'For more information, please visit \n'
+                'https://pyomo.readthedocs.io/en/stable/contributed_packages/mindtpy.html')
+        # config.logger.info("""
+        # If you use this software, you may cite the following:
```
I would edit this reference to the one from PSE2018 until the next work comes out

ZedongPeng replied:
No problem.

bernalde commented on pyomo/contrib/mindtpy/extended_cutting_plane.py:
```diff
@@ -0,0 +1,256 @@
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+"""Implementation of the MindtPy solver.
```
Should this be included in this file? Shouldn't we rather just include it in the main MindtPy file?

bernalde replied:
This seems to be the main file, but why is it called `extended_cutting_plane.py`?

ZedongPeng replied:
The main file is `MindtPy.py`. This file only includes ECP method. I have removed the version descriptions.

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -13,14 +13,12 @@
+def fp_converged(working_model, mip_model, config, discrete_only=True):
     """Calculates the euclidean norm between the discrete variables in the MIP and NLP models.
 
     Parameters
```
This is missing documentation for the newly added parameter `working_model`

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -39,288 +37,63 @@ def fp_converged(solve_data, config, discrete_only=True):
 
 
-def generate_norm_constraint(fp_nlp, solve_data, config):
+def generate_norm_constraint(fp_nlp, mip, config):
```
Just to be consistent with the functions above, let's call all the models in the parameters as `..._model`

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump_new.py:
```diff
@@ -0,0 +1,419 @@
+- Add support for cplex_persistent solver.
+- Fix bug in OA cut expression in cut_generation.py.
+
+TODO: test_FP_OA_8PP will fail. Need to fix.
```
Why?

ZedongPeng replied:
Resolved.

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump_new.py:
```diff
@@ -0,0 +1,419 @@
+                    'MindtPy unable to handle NLP subproblem termination '
+                    'condition of {}'.format(fp_nlp_result.solver.termination_condition))
+            # Call the NLP post-solve callback
+            # TODO fix bug
```
What do you mean with this bug?

ZedongPeng replied:
I don't remember. This might be temporary comment, which I have already resolved.

bernalde commented on pyomo/contrib/mindtpy/nlp_solve.py:
```diff
@@ -95,7 +97,9 @@ def solve_subproblem(solve_data, config):
     nlp_args = dict(config.nlp_solver_args)
-    set_solver_options(nlpopt, solve_data, config, solver_type='nlp')
+    # TODO: Can we move set_solver_options outside of this function?
+    # if not, we can define this function as a method
```
I don't have any strong feelings for one or the other. Wha it's your preference?

ZedongPeng replied:
Since set_solver_options is used in single tree implementation, we should keep it as a function.

bernalde commented on pyomo/contrib/mindtpy/nlp_solve.py:
```diff
@@ -146,6 +151,7 @@ def handle_nlp_subproblem_tc(fixed_nlp, result, solve_data, config, cb_opt=None)
 
 
 def handle_subproblem_optimal(fixed_nlp, solve_data, config, cb_opt=None, fp=False):
+    # TODO: move this to algorithm base class.
```
Yes

ZedongPeng replied:
We already have a `handle_subproblem_optimal` method in the algorithm base class. However, we still need this in the single tree implementation.

bernalde commented on pyomo/contrib/mindtpy/nlp_solve.py:
```diff
@@ -108,6 +112,7 @@ def solve_subproblem(solve_data, config):
 
 
 def handle_nlp_subproblem_tc(fixed_nlp, result, solve_data, config, cb_opt=None):
+    # TODO: move this to algorithm base class.
```
Yes

bernalde commented on pyomo/contrib/mindtpy/tests/unit_test.py:
```diff
@@ -64,266 +60,227 @@ def test_handle_termination_condition(self):
-            solve_data.dual_bound_progress_time = [1, 2, 3, 4, 5, 6]
-            solve_data.dual_bound = 5
-            self.assertEqual(get_dual_integral(solve_data, config), 14.1)
+            # MindtPy = solve_data.working_model.MindtPy_utils
```
Are these commented out because they fail or do they need to be adapted to the new code?

ZedongPeng replied:
The unit test are used for functions. Now the functions becomes methods in the algorithm base class. So it doesn't work now.

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -42,6 +42,7 @@ class MindtPySolveData(object):
     """
     pass
 
+# TODO: move all functions in this file to algorithm base class and solve_data data can be removed.
```
Yes

ZedongPeng replied:
Maybe not all functions. The functions that applies to all strategys and do not depend on solve_data can be kept here.

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -1066,12 +1072,12 @@ def build_ordered_component_lists(model, solve_data):
-    res.problem.name = solve_data.original_model.name
+    res.problem.name = model.name
     res.problem.number_of_nonzeros = None  # TODO
     # TODO work on termination condition and message
```
Maybe this is the right moment to address this TODO

ZedongPeng replied:
Done

bernalde reviewed (commented):
Good first pass @ZedongPeng . We need the code to pass all tests and address all the TODOs that you left behind, but it is looking way better. I like that we are doing this step by step.
I would like to have Can and Kaiyu look into this as soon as they can such that we can have several rounds of reviews here.

Robbybp committed:
clean up imports in implicit_functions.py

Robbybp committed:
trailing comma in import

Robbybp committed:
remove more unusd imports

blnicho committed:
Merge pull request #2892 from Robbybp/implicit-function-imports

Clean up implicit function imports

jsiirola committed:
Add module import tester

jsiirola committed:
Add mpi4py to common.dependencies

jsiirola committed:
Use attemot_import when importing cplex

jsiirola committed:
Update / standardize Qt imports

jsiirola committed:
Correct declaration of raw strings

jsiirola committed:
Leverage attempt_import when importing spreadsheet interface from pyutilib

jsiirola committed:
Remove use of deprecated importlib API

jsiirola committed:
Resolve error deallocating Mumps interface when mumps is not available

jsiirola committed:
Track move of pyomo.core.util module

jsiirola committed:
Avoid deprecation warning by importing sputils from new (private) location

This is a stop-gap to resolve import test failures until PyNumero can be
properly updated (see #2894)

jsiirola committed:
Fix for example that uses the simple_constraint_rule decorator as a function

jsiirola committed:
Update fix_disjuncts so that it can be imported without pyomo.environ

jsiirola committed:
Update tester.py so that it is importable; TODO: remove this file

jsiirola committed:
Add importtest to the singletest job; reorganize jobs to reduce diffs

jsiirola committed:
NFC: commenting diff between workflows

jsiirola committed:
Fix syntax error

jsiirola committed:
Fix syntax error

jsiirola committed:
Apply black

michaelbynum committed:
fix deprecation warnings in pynumero

michaelbynum committed:
Merge branch 'main' into empty_blocks

michaelbynum committed:
test mpi block vector with blocks of size 0

michaelbynum committed:
test mpi block vector with blocks of size 0

michaelbynum committed:
test mpi block vector with blocks of size 0

jsiirola committed:
Reversing quotation characters

jsiirola committed:
Avoid intermediate bash variable

jsiirola committed:
Avoid exception when cplexamp is not available

jsiirola committed:
Development script should be importable

jsiirola committed:
Merge pull request #2895 from michaelbynum/pynumero_deprecations

Remove old, unused PyNumero code

jsiirola committed:
Revert "Avoid deprecation warning by importing sputils from new (private) location"

This reverts commit 6313f2c6a833d5eb0aecc5f2b1d0f7de0b93398d.

jsiirola committed:
Merge branch 'main' into import-tester

michaelbynum committed:
simplify and improve processing of objective bounds from Gurobi

michaelbynum committed:
tests for processing objective bounds from Gurobi for nonconvex QCPs

michaelbynum committed:
run black

ZedongPeng committed:
fix mindtpy singletree implementation bug

ZedongPeng committed:
add description for mip_start_lazy_oa_cuts

michaelbynum committed:
Merge pull request #2898 from michaelbynum/appsi_gurobi_results

Correctly Process Objective Bounds from Gurobi for Nonconvex QCPs

ZedongPeng committed:
add comments

ZedongPeng committed:
format

Robbybp committed:
add PyNumeroEvaluationError and use it in ASL interface instead of asserting return values

Robbybp committed:
add blank line for black

michaelbynum committed:
change default behavior in appsi to not care where variables live

michaelbynum committed:
run black

jsiirola committed:
Move TemplateExpressionError to common.errros

jsiirola committed:
Move value(), check_if_numeric_type() to common.numeric_types

jsiirola committed:
Move NumericValue, NumericNDArray to expr.numeric_expr

jsiirola committed:
Break import loop by deferring the visitor import

jsiirola committed:
Silence (unnecessary) test output to console

jsiirola committed:
Remove private classes/functions from expr.current

jsiirola committed:
Move numerif function defitions into expr.numeric_expr

michaelbynum committed:
improve appsi tests and fix a bug

michaelbynum committed:
run black

Commit:
Adding function for collecting all expressions with unit inconsistencies

ZedongPeng committed:
change the name of feasibility pump

ZedongPeng committed:
remove unused import

ZedongPeng committed:
simplify process_objective function args

Robbybp committed:
add tests catching PyNumeroEvaluationError

ZedongPeng committed:
add exception_flag=False for solver.available()

jsiirola committed:
Define UnavailableClass to generate base classes that raise DeferredImportError upon instantiation

jsiirola committed:
Improve error message when using unavailable class

jsiirola committed:
Track removal of bool from native_numeric_types

jsiirola committed:
Update UnavailableClass to raise DeferredImportError on attribute access

jsiirola committed:
NFC: fix typo

jsiirola committed:
Merge branch 'main' into import-tester

jsiirola committed:
Move expression imports from expr.current to core.expr

jsiirola committed:
Remove references to pyomo.core.expr.current

jsiirola committed:
deprecate pyomo.core.expr.current

jsiirola committed:
Merge pull request #2897 from jsiirola/import-tester

Ensure all pyomo modules import without warnings

jsiirola committed:
Track changes to expr.Mode

jsiirola committed:
Apply black

jsiirola committed:
Merge branch 'main' into reorg-expr-imports

jsiirola committed:
Merge remote-tracking branch 'refs/remotes/jsiirola/reorg-expr-imports' into reorg-expr-imports

jsiirola committed:
Merge branch 'reorg-expr-imports' into remove-expr-current

jsiirola committed:
Track change in expr.Mode

jsiirola committed:
Merge branch 'reorg-expr-imports' into remove-expr-current

jsiirola committed:
Fix compatibility import

jsiirola committed:
Merge branch 'reorg-expr-imports' into remove-expr-current

jsiirola committed:
Update Expr_if to support Cythonization

jsiirola committed:
Avoid accessing locals() when processing Expr_if args

ZedongPeng committed:
remove config in args

michaelbynum committed:
Merge pull request #2901 from Robbybp/pynumero-eval-error

Add `PyNumeroEvaluationError`

jsiirola committed:
Merge branch 'main' into remove-expr-current

emma58 committed:
Merge branch 'main' into appsi_update

jsiirola committed:
Resolve exception in FileDownloader.get_os_version when lsb_release is missing

Commit:
Fixing grammatical error

Commit:
Merge branch 'main' into inconsistent_units

mrmundt committed:
Add in expected failuress for gams + MILP_unbounded

mrmundt committed:
Update GAMS test version

mrmundt committed:
Pin back the GAMS version and add a note

jsiirola committed:
document UnavailableClass

jsiirola committed:
document format_exception()

jsiirola committed:
move DeferredImportError to pyomo.common.errors

jsiirola committed:
Alphabetize pyomo/common/errors.py

jsiirola committed:
Add pyomo.common.errors to the library reference

jsiirola committed:
Clairify common documentation; update links

jsiirola committed:
Make the SymbolMap used by the GAMS writer picklable

jsiirola committed:
Apply black

mrmundt committed:
HACK: Manually mark the cases we want to skip

jsiirola committed:
Add variable, constraint names to exception messages

mrmundt committed:
HACK: Horrible hack but picks up the right tests

jsiirola committed:
NFC: fix typo

jsiirola committed:
Improve errors raised by calculate_variable_from_constraint

mrmundt committed:
Merge pull request #2913 from jsiirola/gams-pickle

GAMS: fix pickling models, update tests

blnicho committed:
NFC: Fix docstring typo in errors.py

emma58 committed:
Fixing a typo in the hull transformation, and adding a test for it

blnicho committed:
Merge branch 'main' into common-docs

blnicho committed:
Merge branch 'main' into downloader-lsb-release

bernalde committed:
Merge branch 'main' into mindtpy-rewrite

jsiirola committed:
Deprecate pyomo.core.kernel.expression.noclone

jsiirola committed:
Always clone 'Floating' components (not attached to blocks)

jsiirola committed:
Resync kernel's __deepcopy__ with core.base's

This also switches kernel to use the same block scope indicator in the
deepcopy memo so that kernel models work better with Pyomo functions
that use the low-level interface to deepcopy.

jsiirola committed:
Resolve expected failures marked in #2913

jsiirola committed:
Remove the 'singletest' job from "branches" builds

ZedongPeng committed:
add warmstart for ipopt

mrmundt committed:
Merge pull request #2916 from jsiirola/deprecate-noclone

Resolve kernel pickle/clone errors

mrmundt committed:
Merge pull request #2915 from emma58/hull-typo-fix

Fixing a typo in gdp.hull transformation

mrmundt committed:
Merge pull request #2911 from jsiirola/downloader-lsb-release

Resolve `FileDownloader.get_os_version` exception for missing `lsb_release`

jsiirola committed:
NFC: fix grammar

jsiirola committed:
Merge branch 'main' into calc-var-value-messages

jsiirola committed:
Apply black

mrmundt committed:
Merge branch 'main' into inconsistent_units

mrmundt committed:
Merge pull request #2912 from jsiirola/common-docs

Update `pyomo.common` documentation

jsiirola committed:
Merge branch 'main' into calc-var-value-messages

jsiirola committed:
Merge pull request #2917 from jsiirola/gha-remote-branches-singletest

Remove the 'singletest' job from "branches" builds

jsiirola committed:
Merge branch 'main' into appsi_update

jsiirola committed:
Merge branch 'main' into reorg-expr-imports

jsiirola committed:
Update import location

jsiirola committed:
NFC: fix typo

jsiirola committed:
Merge pull request #2909 from jsiirola/reorg-expr-imports

Reorganize numeric expression modules

jsiirola committed:
Merge branch 'main' into remove-expr-current

jsiirola committed:
Merge pull request #2914 from jsiirola/calc-var-value-messages

Improve `calculate_variable_from_constraint` error messages

jsiirola committed:
NFC: fix typo

jsiirola committed:
Remove duplicate definition

jsiirola committed:
Merge branch 'main' into mindtpy-rewrite

jsiirola committed:
Apply black

jsiirola committed:
Xpress: only fall back on the LP results processor if there is no NLP status

jsiirola committed:
Apply black

jsiirola committed:
NFX: fix typos

jsiirola committed:
Merge pull request #2921 from jsiirola/xpress-nlp-status-fix

Xpress: fix NLP results processing for Xpress 9.2.0

jsiirola committed:
Merge branch 'main' into remove-expr-current

ZedongPeng committed:
fix lp/nlp bug

ZedongPeng committed:
Merge branch 'mindtpy-rewrite' of github.com:ZedongPeng/pyomo into mindtpy-rewrite

ZedongPeng committed:
set update_config for feasibility_nlp_opt

ZedongPeng committed:
rename add_feas_slacks to initialize_feas_subproblem

ZedongPeng committed:
Merge branch 'main' into mindtpy-rewrite

Commit:
Adding logger messages

Commit:
Merge branch 'inconsistent_units' of https://github.com/andrewlee94/pyomo into inconsistent_units

jsiirola committed:
Cleaning up imports, per review

marvin-meck committed:
pass dim=0 to _process_data_list if scalar param is encounterd, fixes issue #120

marvin-meck committed:
updated ./pyomo/dataportal/tests/test_dataportal.py to catch bug reported in issue #120

Commit:
Fix model viewer

jsiirola committed:
Merge pull request #2904 from andrewlee94/inconsistent_units

Adding function for collecting all expressions with unit inconsistencies

marvin-meck committed:
autoformatted ./pyomo/dataportal/test_dataportal.py

marvin-meck committed:
Merge branch 'main' into bugfix-no120

Commit:
Fix model seletor

Commit:
Run black

Commit:
Merge branch 'main' of https://github.com/pyomo/pyomo into fix_model_viewer

mrmundt committed:
Merge pull request #2910 from jsiirola/remove-expr-current

Remove `expr.current`

Commit:
Fix left over debug code

Commit:
Merge branch 'main' of https://github.com/pyomo/pyomo into fix_model_viewer

jsiirola committed:
Merge pull request #2923 from marvin-meck/bugfix-no120

BUGFIX: Loading Scalar Param from `.dat` file

ZedongPeng committed:
change MILP to MIP

ZedongPeng committed:
change MIQP to MIQCP

ZedongPeng committed:
change MILP to MIP

ZedongPeng committed:
fix typo

mrmundt committed:
Typos version update

blnicho committed:
Merge pull request #2931 from mrmundt/typos-update

Typos version update

blnicho committed:
Merge pull request #2924 from eslickj/fix_model_viewer

Fix model viewer

ZedongPeng committed:
change MILP to MIP

ZedongPeng committed:
remove TODO

ZedongPeng committed:
change todo to comment

ZedongPeng committed:
resolve tabu list todo

ZedongPeng committed:
change the name of simpleMINLP example

ZedongPeng committed:
add special handle with mip start solution in single tree

ZedongPeng committed:
add initial value for constraint qualification example

ZedongPeng committed:
use model.clone() in mindtpy test

ZedongPeng committed:
add special handle of initial binary strategy

ZedongPeng committed:
Merge branch 'mindtpy-rewrite' of github.com:ZedongPeng/pyomo into mindtpy-rewrite

ZedongPeng committed:
black format

ZedongPeng committed:
Merge branch 'main' into mindtpy-rewrite

michaelbynum committed:
Merge branch 'main' into appsi_update

michaelbynum committed:
update comment

michaelbynum committed:
Merge remote-tracking branch 'michaelbynum/appsi_update' into appsi_update

michaelbynum committed:
merge main into empty_blocks

michaelbynum committed:
tests

michaelbynum committed:
Merge pull request #2903 from michaelbynum/appsi_update

change default behavior in appsi to not care where variables live

blnicho committed:
Improve error message when attempting to discretize models with circular references

michaelbynum committed:
fix test

ZedongPeng committed:
add description for mip_strategy_presolvenode

ZedongPeng committed:
remove handle_main_other_conditions function

ZedongPeng committed:
change wrt to w.r.t.

ZedongPeng committed:
fix typo

ZedongPeng committed:
fix bug

ZedongPeng committed:
fix typo

ZedongPeng committed:
change [...] to values()

ZedongPeng committed:
fix typo

mrmundt committed:
Merge pull request #2943 from blnicho/dae-component-names

DAE: Improve error message when attempting to discretize models with circular references

ZedongPeng committed:
add comments for callback

ZedongPeng committed:
improve comments

ZedongPeng committed:
fix mip_start_lazy_oa_cuts bug

ZedongPeng committed:
remove clone of ConstraintQualificationExample

ZedongPeng committed:
change the way of checking empty list

ZedongPeng committed:
update the link of CPLEX

ZedongPeng committed:
change number to enum

ZedongPeng committed:
add the log of error message

ZedongPeng committed:
Merge branch 'main' into mindtpy-rewrite

ZedongPeng committed:
black format

ZedongPeng committed:
improve constraint list and var list generation

mrmundt committed:
Update PR template to add linting instructions

mrmundt committed:
Adjust formatting to avoid wraparound

mrmundt committed:
Last change didn't commit?

mrmundt committed:
Merge pull request #2947 from mrmundt/main

Add Explicit Linting Instructions to PR Template

bernalde committed:
Merge branch 'main' into mindtpy-rewrite

ZedongPeng committed:
rename mindtpy_object to mindtpy_solver

ZedongPeng committed:
rename set_solver_timelimit to update_solver_timelimit

ZedongPeng committed:
change remaining time calculation from int to ceil

ZedongPeng committed:
Merge branch 'mindtpy-rewrite' of github.com:ZedongPeng/pyomo into mindtpy-rewrite

ZedongPeng committed:
black format

mrmundt committed:
Merge branch 'main' into empty_blocks

jsiirola committed:
Switch named expressions to store in

jsiirola committed:
Names expressions: `expr` is guaranteed to be a NumericValue

jsiirola committed:
Define __call__ to trap unconstructed ScalarExpression/ScalarObjective

jsiirola committed:
DRY: leverage _pop_from_kwargs()

jsiirola committed:
FBBT: access named expressions through `args`

jsiirola committed:
Apply black

jsiirola committed:
Merge pull request #2896 from michaelbynum/empty_blocks

Make min and max work with MPIBlockVector when some blocks have size 0

mrmundt committed:
Remove HTML4 forcing

emma58 committed:
Fixing a that vs. which mistake in an error message

jsiirola committed:
Resolve error categorizing some discrete domains as integer domains

mrmundt committed:
Merge pull request #2950 from emma58/grammar-nit

Fixing a 'that' vs. 'which' mistake in error about bounds crossing

mrmundt committed:
Merge pull request #2949 from mrmundt/docs-theme-update

Remove HTML4 Forcing in Online Docs

mrmundt committed:
Merge pull request #2954 from jsiirola/var-is-integer

Resolve error categorizing some Var discrete domains as "integer"

jsiirola committed:
NFC: cleaning up exception messages

jsiirola committed:
Merge branch 'main' into named-expr-is-numericvalue

jsiirola committed:
Support kwargs in partial objects passed to Initializer

mrmundt committed:
Merge pull request #2948 from jsiirola/named-expr-is-numericvalue

Named expressions: `expr` should always return `NumericValue`

jsiirola committed:
Merge pull request #2960 from jsiirola/initializer-partial=kwargs

Support kwargs in partial objects passed to Initializer()

mrmundt committed:
Merge branch 'main' into mindtpy-rewrite

ZedongPeng committed:
fix math domain error

ZedongPeng committed:
fix docstring

ZedongPeng committed:
change GUROBI to Gurobi

ZedongPeng committed:
remove appsi_highs support

ZedongPeng committed:
remove wrong calculate_dual_at_solution check

ZedongPeng committed:
black format

ZedongPeng committed:
add comments

ZedongPeng committed:
add comments for optimalitytarget

ZedongPeng committed:
black format
